sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs

```diff
@@ -0,0 +1,473 @@
+            let connection = self.ensure_connection().await?;
+
+            let session = AmqpSession::new();
+            session
+                .begin(
+                    connection.as_ref(),
+                    Some(AmqpSessionOptions {
+                        incoming_window: Some(u32::MAX),
+                        outgoing_window: Some(u32::MAX),
+                        ..Default::default()
+                    }),
+                )
+                .await?;
+            session_instances.insert(source_url.clone(), Arc::new(session));
+        }
+        let rv = session_instances
+            .get(source_url)
+            .ok_or_else(|| EventHubsError::from(ErrorKind::MissingSession))?
+            .clone();
+        debug!("Cloning session for partition {:?}", source_url);
+        Ok(rv)
+    }
+
+    async fn create_connection(&self) -> Result<Arc<AmqpConnection>> {
+        trace!("Creating connection for {}.", self.url);
+        let connection = Arc::new(AmqpConnection::new());
+
+        connection
+            .open(
+                self.connection_name.clone(),
+                self.url.clone(),
+                Some(AmqpConnectionOptions {
+                    properties: Some(
+                        vec![
+                            ("user-agent", get_user_agent(&self.application_id)),
+                            ("version", get_package_version()),
+                            ("platform", get_platform_info()),
+                            ("product", get_package_name()),
+                        ]
+                        .into_iter()
+                        .map(|(k, v)| (AmqpSymbol::from(k), AmqpValue::from(v)))
+                        .collect(),
+                    ),
+                    custom_endpoint: self.custom_endpoint.clone(),
+                    ..Default::default()
+                }),
+            )
+            .await?;
+        Ok(connection)
+    }
+
+    pub(super) async fn ensure_amqp_management(self: &Arc<Self>) -> Result<Arc<AmqpManagement>> {
+        let management_client = self
+            .mgmt_client
+            .get_or_try_init(|| {
+                RecoverableManagementClient::create_management_client(
+                    self.clone(),
+                    &self.retry_options,
+                )
+            })
+            .await?;
+
+        Ok(management_client.clone())
+    }
+
+    /// Ensures that the AMQP Claims-Based Security (CBS) client is created and attached.
+    pub(super) async fn ensure_amqp_cbs(self: &Arc<Self>) -> Result<Arc<AmqpClaimsBasedSecurity>> {
+        debug!("Ensuring AMQP Claims-Based Security (CBS) client.");
+
+        let connection = self.ensure_connection().await?;
+        let cbs_client = RecoverableClaimsBasedSecurity::create_claims_based_security(
+            connection.clone(),
+            &self.retry_options,
+        )
+        .await?;
+        debug!("AMQP Claims-Based Security (CBS) client ensured.");
```

Comment:
Would a span make more sense here? It would automatically have begin/end and encapsulate any child spans/logs. Just a thought.

Reply:
Probably not a bad idea, tbh.